fs/localfs/copy.go

```diff
@@ -0,0 +1,164 @@
+}
+
+// set permission, modification time and user/group ids on targetPath
+func setAttributes(targetPath string, e fs.Entry) error {
+	const modBits = os.ModePerm | os.ModeSetgid | os.ModeSetuid | os.ModeSticky
+
+	le, err := NewEntry(targetPath)
+	if err != nil {
+		return errors.Wrap(err, "could not create local FS entry for "+targetPath)
+	}
+
+	// Set owner user and group from e
+	if le.Owner() != e.Owner() {
+		if err = os.Chown(targetPath, int(e.Owner().UserID), int(e.Owner().GroupID)); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change owner/group for "+targetPath)
+		}
+	}
+
+	// Set file permissions from e
+	if (le.Mode() & modBits) != (e.Mode() & modBits) {
+		if err = os.Chmod(targetPath, e.Mode()&modBits); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change permissions on "+targetPath)
+		}
+	}
+
+	// Set mod time from e
+	if !le.ModTime().Equal(e.ModTime()) {
+		// Note: Set atime to ModTime as well
+		if err = os.Chtimes(targetPath, e.ModTime(), e.ModTime()); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change mod time on "+targetPath)
+		}
+	}
+
+	return nil
+}
+
+func copyDirectory(ctx context.Context, d fs.Directory, targetPath string) error {
+	if err := createDirectory(targetPath); err != nil {
+		return err
+	}
+
+	return copyDirectoryContent(ctx, d, targetPath)
+}
+
+func copyDirectoryContent(ctx context.Context, d fs.Directory, targetPath string) error {
+	entries, err := d.Readdir(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range entries {
+		if err := copyEntry(ctx, e, filepath.Join(targetPath, e.Name())); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func createDirectory(path string) error {
+	switch stat, err := os.Stat(path); {
+	case os.IsNotExist(err):
+		return os.MkdirAll(path, 0700)
+	case err != nil:
+		return errors.Wrap(err, "failed to stat path "+path)
+	case stat.Mode().IsDir():
+		return errors.Errorf("directory already exists, not overwriting it: %q", path)
+	default:
+		return errors.Errorf("unable to create directory, %q already exists and it is not a directory", path)
+	}
+}
+
+func copyFileContent(ctx context.Context, targetPath string, f fs.File) error {
+	var out *os.File
+
+	switch _, err := os.Stat(targetPath); {
+	case os.IsNotExist(err):
+		if out, err = os.Create(targetPath); err != nil {
+			return errors.Wrap(err, "failed to create new file "+targetPath)
+		}
+		defer out.Close() //nolint:errcheck
+	case err == nil:
+		return errors.Errorf("unable to create %q, it already exists", targetPath)
+	default:
+		return errors.Wrap(err, "failed to stat "+targetPath)
+	}
+
+	r, err := f.Open(ctx)
+	if err != nil {
+		return errors.Wrap(err, "unable to open snapshot file for "+targetPath)
+	}
+	defer r.Close() //nolint:errcheck
+
+	if _, err = io.Copy(out, r); err != nil {
+		return errors.Wrapf(err, "unable to restore %q content", targetPath)
+	}
+
+	return out.Sync()
```

Comment:
do we need to sync here? That might kill performance of restores.

Reply:
Mmmm, we may be able to remove it. 

* Yes, it may kill performance. OTOH, `r.Close()` may (or may not) trigger a sync, and whether or not that ends up being synchronous is OS/FS dependent. This is not explicitly specified in `close()` for POSIX, and most implementations end up being asynchronous, and thus the application has no way of knowing whether or not the data was persisted.
* The intent here was to capture write errors and surface those as soon as possible. So, this would be the _safest_ option. At the same time, it seems reasonable that the default behavior could be to *NOT sync*, and have a flag to overwrite that.

I'll go ahead and remove the `Sync()` and at least capture the error from `out.Close()` when there have been no other errors.

Thoughts?

Reply:
Removed the `Sync()` when leveraging `atomic.WriteFile(...)`